From 4a04f900e167bbf3aecee89875fc13cd2112df38 Mon Sep 17 00:00:00 2001
From: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Date: Mon, 26 Jul 2021 19:14:44 +0300
Subject: [PATCH 3/4] device-tree: Introduce scmi nodes to work with clocks and
 resets via scmi

Add separate file with the description of the scmi nodes.
Update dt nodes by setting clocks, resets and power-domains to
be operated through scmi.
Changes were done:
 + Add firmware/scmi node with the description of
 clock/reset/power-domain subnodes;
 * Set the following devices to be operated via scmi:
    - avb (ethernet);
    - hdmi0;
    - i2c2, i2c4;
    - usb_dmac{0,1,2,3};
    - hsusb
    - xhci, ehci, ohci devices;
    - audma{0,1};
    - rcar_sound;

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Signed-off-by: Sergiy Kibrik <Sergiy_Kibrik@epam.com>
Reviewed-by:  Sergiy Kibrik <Sergiy_Kibrik@epam.com>
Reviewed-by: Andrii Anisov <andrii_anisov@epam.com>
---
 .../arm64/boot/dts/renesas/r8a77951-scmi.dtsi | 238 ++++++++++++++++++
 arch/arm64/boot/dts/renesas/r8a77951-ulcb.dts |   1 +
 2 files changed, 239 insertions(+)
 create mode 100644 arch/arm64/boot/dts/renesas/r8a77951-scmi.dtsi

diff --git a/arch/arm64/boot/dts/renesas/r8a77951-scmi.dtsi b/arch/arm64/boot/dts/renesas/r8a77951-scmi.dtsi
new file mode 100644
index 000000000000..00e22e6942f7
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a77951-scmi.dtsi
@@ -0,0 +1,238 @@
+/*
+ * Device Tree scmi configuration.
+ *
+ * Copyright (C) 2021 EPAM Systems.
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+/ {
+	cpu_scp_shm: scp-shmem@0x53FF0000 {
+		compatible = "arm,scmi-shmem";
+		reg = <0x0 0x53FF0000 0x0 0x1000>;
+	};
+
+	firmware {
+		scmi {
+			compatible = "arm,scmi-smc";
+			arm,smc-id = <0x82000002>;
+			shmem = <&cpu_scp_shm>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			scmi_power: protocol@11 {
+				reg = <0x11>;
+				#power-domain-cells = <1>;
+			};
+
+			scmi_clock: protocol@14 {
+				reg = <0x14>;
+				#clock-cells = <1>;
+			};
+
+			scmi_reset: protocol@16 {
+				reg = <0x16>;
+				#reset-cells = <1>;
+			};
+		};
+	};
+
+	reserved-memory {
+		/* reserved region for scmi channels*/
+		scmi_memory: linux,scmi_mem@53FF0000 {
+			no-map;
+			reg = <0x0 0x53FF0000 0x0 0x10000>;
+		};
+	};
+};
+
+&soc {
+	i2c2_clk: i2c2_clk {
+		compatible = "fixed-clock";
+		/*
+		The value "133333328" was taken from /sys/kernel/debug/clk/clk_summary
+		for the corresponding clock in DomD.
+		*/
+		clock-frequency = <133333328>;
+		#clock-cells = <0>;
+	};
+};
+
+&avb {
+	scmi_devid = <0>;
+	clocks = <&scmi_clock 0>;
+	power-domains = <&scmi_power 0>;
+	resets = <&scmi_reset 0>;
+};
+
+&hdmi0 {
+	scmi_devid = <1>;
+	resets = <&scmi_reset 1>;
+};
+
+&i2c4 {
+	scmi_devid = <7>;
+	resets = <&scmi_reset 7>;
+};
+
+&i2c2 /* i2c@e6510000*/
+{
+	scmi_devid = <5>;
+	resets = <&scmi_reset 5>;
+};
+
+&usb_dmac2 /* dma-controller@e6460000*/
+{
+	scmi_devid = <15>;
+	clocks = <&scmi_clock 7>;
+	resets = <&scmi_reset 14>;
+	power-domains = <&scmi_power 0>;
+};
+
+&usb_dmac3 /* dma-controller@e6470000*/
+{
+	scmi_devid = <16>;
+	clocks = <&scmi_clock 8>;
+	resets = <&scmi_reset 15>;
+	power-domains = <&scmi_power 0>;
+};
+
+&usb_dmac0 /* dma-controller@e65a0000*/
+{
+	scmi_devid = <13>;
+	clocks = <&scmi_clock 5>;
+	resets = <&scmi_reset 12>;
+	power-domains = <&scmi_power 0>;
+};
+
+&usb_dmac1 /* dma-controller@e65b0000*/
+{
+	scmi_devid = <14>;
+	clocks = <&scmi_clock 6>;
+	resets = <&scmi_reset 13>;
+	power-domains = <&scmi_power 0>;
+};
+
+&hsusb /* usb@e6590000*/
+{
+	scmi_devid = <19>;
+	clocks = <&scmi_clock 3>, <&scmi_clock 2>;
+	resets = <&scmi_reset 10>, <&scmi_reset 9>;
+	power-domains = <&scmi_power 0>;
+};
+
+&xhci0 /* usb@ee000000*/
+{
+	scmi_devid = <8>;
+	clocks = <&scmi_clock 1>;
+	resets = <&scmi_reset 8>;
+	power-domains = <&scmi_power 0>;
+};
+
+&ohci0 /* usb@ee080000*/
+{
+	scmi_devid = <9>;
+	clocks = <&scmi_clock 2>, <&scmi_clock 3>;
+	resets = <&scmi_reset 9>, <&scmi_reset 10>;
+	power-domains = <&scmi_power 0>;
+};
+
+&ehci0 /* usb@ee080100*/
+{
+	scmi_devid = <10>;
+	clocks = <&scmi_clock 2>, <&scmi_clock 3>;
+	resets = <&scmi_reset 9>, <&scmi_reset 10>;
+	power-domains = <&scmi_power 0>;
+};
+
+&ohci1 /* usb@ee0a0000*/
+{
+	scmi_devid = <11>;
+	clocks = <&scmi_clock 4>;
+	resets = <&scmi_reset 11>;
+	power-domains = <&scmi_power 0>;
+};
+
+&ehci1 /* usb@ee0a0100*/
+{
+	scmi_devid = <12>;
+	clocks = <&scmi_clock 4>;
+	resets = <&scmi_reset 11>;
+	power-domains = <&scmi_power 0>;
+};
+
+&audma0 /* dma-controller@ec700000*/
+{
+	scmi_devid = <21>;
+	clocks = <&scmi_clock 36>;
+	resets = <&scmi_reset 28>;
+	power-domains = <&scmi_power 0>;
+};
+
+&audma1 /* dma-controller@ec720000*/
+{
+	scmi_devid = <22>;
+	clocks = <&scmi_clock 35>;
+	resets = <&scmi_reset 27>;
+	power-domains = <&scmi_power 0>;
+};
+
+&rcar_sound /* sound@ec500000*/
+{
+	scmi_devid = <20>;
+
+	resets = <&scmi_reset 16>,
+		 <&scmi_reset 17>, <&scmi_reset 18>,
+		 <&scmi_reset 19>, <&scmi_reset 20>,
+		 <&scmi_reset 21>, <&scmi_reset 22>,
+		 <&scmi_reset 23>, <&scmi_reset 24>,
+		 <&scmi_reset 25>, <&scmi_reset 26>;
+
+	clocks =
+			 /* "ssi-all" */
+			 <&scmi_clock 9>,
+			 /* "ssi.9", "ssi.8" */
+			 <&scmi_clock 10>, <&scmi_clock 11>,
+			 /* "ssi.7", "ssi.6" */
+			 <&scmi_clock 12>, <&scmi_clock 13>,
+			 /* "ssi.5", "ssi.4" */
+			 <&scmi_clock 14>, <&scmi_clock 15>,
+			 /* "ssi.3", "ssi.2" */
+			 <&scmi_clock 16>, <&scmi_clock 17>,
+			 /* "ssi.1", "ssi.0" */
+			 <&scmi_clock 18>, <&scmi_clock 19>,
+			 /* "src.9", "src.8" */
+			 <&scmi_clock 25>, <&scmi_clock 26>,
+			 /* "src.7", "src.6" */
+			 <&scmi_clock 27>, <&scmi_clock 28>,
+			 /* "src.5", "src.4" */
+			 <&scmi_clock 29>, <&scmi_clock 30>,
+			 /* "src.3", "src.2" */
+			 <&scmi_clock 31>, <&scmi_clock 32>,
+			 /* "src.1", "src.0" */
+			 <&scmi_clock 33>, <&scmi_clock 34>,
+			 /* "mix.1", "mix.0" */
+			 <&scmi_clock 23>, <&scmi_clock 24>,
+			 /* "ctu.1", "ctu.0" */
+			 <&scmi_clock 23>, <&scmi_clock 24>,
+			 /* "dvc.0", "dvc.1" */
+			 <&scmi_clock 22>, <&scmi_clock 21>,
+			 /* "clk_a", "clk_b" */
+			 <&audio_clk_a>, <&cs2000>,
+			 /* "clk_c", "clk_i" */
+			 <&audio_clk_c>, <&i2c2_clk>;
+
+		clock-names = "ssi-all",
+				  "ssi.9", "ssi.8", "ssi.7", "ssi.6",
+				  "ssi.5", "ssi.4", "ssi.3", "ssi.2",
+				  "ssi.1", "ssi.0",
+				  "src.9", "src.8", "src.7", "src.6",
+				  "src.5", "src.4", "src.3", "src.2",
+				  "src.1", "src.0",
+				  "mix.1", "mix.0",
+				  "ctu.1", "ctu.0",
+				  "dvc.0", "dvc.1",
+				  "clk_a", "clk_b", "clk_c", "clk_i";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a77951-ulcb.dts b/arch/arm64/boot/dts/renesas/r8a77951-ulcb.dts
index fe4c0396fab4..32daf541f898 100644
--- a/arch/arm64/boot/dts/renesas/r8a77951-ulcb.dts
+++ b/arch/arm64/boot/dts/renesas/r8a77951-ulcb.dts
@@ -9,6 +9,7 @@
 /dts-v1/;
 #include "r8a77951.dtsi"
 #include "ulcb.dtsi"
+#include "r8a77951-scmi.dtsi"
 
 / {
 	model = "Renesas H3ULCB board based on r8a77951";
-- 
2.25.1

